api/src/wfl/sink.clj

```diff
@@ -242,21 +243,105 @@
+                            {:status status :updated (utc-now)}
+                            ["id = ?" job-id])))]
+    (doseq [[id job-id workflow] (read-active-jobs)]
+      (let [job    (datarepo/job-metadata job-id)
+            status (-> job :job_status str/lower-case)]
+        (write-job-status id status)
+        (when (= "failed" status)
+          (throw (UserException. "A DataRepo ingest job failed"
+                                 {:job      job
+                                  :workflow workflow})))))))
+
+(defn ^:private to-dataset-row
+  "Use `fromOutputs` to coerce the `workflow` outputs into a row in the dataset
+   where `fromOutputs` describes a mapping from workflow outputs to columns in
+   the dataset table."
+  [fromOutputs {:keys [outputs] :as workflow}]
+  (when-not (map? fromOutputs)
+    (throw (IllegalStateException. "fromOutputs is malformed")))
+  (try
+    (rename-gather outputs fromOutputs)
+    (catch Exception cause
+      (throw (ex-info "Failed to coerce workflow outputs to dataset columns"
+                      {:fromOutputs fromOutputs :workflow workflow}
+                      cause)))))
+
+(defn ^:private start-ingesting-outputs
+  "Start ingesting the `workflow` outputs as a new row in the target dataset
+   and push the tdr ingest job into the `sink`'s job queue."
+  [{:keys [dataset table fromOutputs] :as sink}
+   {:keys [uuid] :as workflow}]
+  (let [file (storage/gs-url "broad-gotc-dev-wfl-ptc-test-outputs" (str uuid ".json"))]
+    (-> (to-dataset-row fromOutputs workflow)
+        (json/write-str :escape-slash false)
+        (storage/upload-content file))
+    (->> (datarepo/ingest-table (:id dataset) file table)
+         (assoc {:status "running" :workflow uuid} :job)
+         (push-job sink))))
+
 (defn ^:private update-datarepo-sink
-  [_executor _sink]
-  (throw (NotImplementedException. "update-datarepo-sink")))
+  "Attempt to a pull a workflow off the upstream `executor` queue and write its
+   outputs as new rows in a tdr dataset table asynchronously."
+  [executor sink]
+  (when-let [[_ workflow] (stage/peek-queue executor)]
+    (start-ingesting-outputs sink workflow)
+    (stage/pop-queue! executor))
+  (update-datarepo-job-statuses sink)
+  (when-let [{:keys [workflow job] :as record} (peek-job-queue sink)]
+    (try
+      (let [res (datarepo/get-job-result job)]
+        (log/info "Sunk workflow outputs to dataset"
+                  {:dataset (get-in sink [:dataset :id]) :workflow workflow}))
+      (finally
+        (pop-job-queue! sink record)))))
 
-(defn ^:private datarepo-sink-done? [{:keys [details] :as _sink}]
+(defn ^:private datarepo-sink-done?
+  "True when all tdr ingest jobs created by the `_sink` have terminated."
+  [{:keys [details] :as _sink}]
   (let [query "SELECT COUNT(*) FROM %s
-               WHERE status   <> 'failed'
-               AND   consumed IS NOT NULL"]
+               WHERE status <> 'failed' AND consumed IS NULL"]
```

Comment:
(Fun with Postgres) if an unconsumed record has null status it won't be included in the count, not sure if that's relevant for this sink impl.

Reply:
Thanks for spotting that! I don't think it can have a null status. It's certainly an error if it does.